riscv-executor/src/large_field/mod.rs

```diff
@@ -0,0 +1,1363 @@
+        }
+    }
+
+    pub fn bin(&self) -> i64 {
+        match self {
+            Self::Binary(b) => *b,
+            Self::Field(_) => panic!(),
+        }
+    }
+
+    fn u(&self) -> u32 {
+        self.bin().try_into().unwrap()
+    }
+
+    fn s(&self) -> i32 {
+        self.bin().try_into().unwrap()
+    }
+
+    fn is_zero(&self) -> bool {
+        match self {
+            Self::Binary(b) => *b == 0,
+            Self::Field(f) => f.is_zero(),
+        }
+    }
+
+    fn add(&self, other: &Self) -> Self {
+        match (self, other) {
+            (Self::Binary(a), Self::Binary(b)) => Self::Binary(a.checked_add(*b).unwrap()),
+            (Self::Field(a), Self::Field(b)) => Self::Field(*a + *b),
+            (Self::Binary(a), Self::Field(b)) => Self::Field(F::from(*a) + *b),
+            (Self::Field(a), Self::Binary(b)) => Self::Field(*a + F::from(*b)),
+        }
+    }
+
+    fn sub(&self, other: &Self) -> Self {
+        match (self, other) {
+            (Self::Binary(a), Self::Binary(b)) => Self::Binary(a.checked_sub(*b).unwrap()),
+            (Self::Field(a), Self::Field(b)) => Self::Field(*a - *b),
+            (Self::Binary(a), Self::Field(b)) => Self::Field(F::from(*a) - *b),
+            (Self::Field(a), Self::Binary(b)) => Self::Field(*a - F::from(*b)),
+        }
+    }
+
+    fn mul(&self, other: &Self) -> Self {
+        match (self, other) {
+            (Self::Binary(a), Self::Binary(b)) => match a.checked_mul(*b) {
+                Some(v) => Self::Binary(v),
+                None => {
+                    let a = F::from(*a);
+                    let b = F::from(*b);
+                    Self::Field(a * b)
+                }
+            },
+            (Self::Field(a), Self::Field(b)) => Self::Field(*a * *b),
+            (Self::Binary(a), Self::Field(b)) => Self::Field(F::from(*a) * *b),
+            (Self::Field(a), Self::Binary(b)) => Self::Field(*a * F::from(*b)),
+        }
+    }
+}
+
+impl<F: FieldElement> From<i64> for Elem<F> {
+    fn from(value: i64) -> Self {
+        Self::Binary(value)
+    }
+}
+
+impl<F: FieldElement> From<u32> for Elem<F> {
+    fn from(value: u32) -> Self {
+        Self::Binary(value as i64)
+    }
+}
+
+impl<F: FieldElement> From<i32> for Elem<F> {
+    fn from(value: i32) -> Self {
+        Self::Binary(value as i64)
+    }
+}
+
+impl<F: FieldElement> From<usize> for Elem<F> {
+    fn from(value: usize) -> Self {
+        Self::Binary(value as i64)
+    }
+}
+
+impl<F: FieldElement> Display for Elem<F> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Binary(b) => write!(f, "{b}"),
+            Self::Field(fe) => write!(f, "{fe}"),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct RegisterMemory<F: FieldElement> {
+    last: HashMap<u32, Elem<F>>,
+    second_last: HashMap<u32, Elem<F>>,
+}
+
+impl<F: FieldElement> RegisterMemory<F> {
+    pub fn for_bootloader(&self) -> HashMap<u32, Vec<F>> {
```

Comment:
The return type here was `HashMap<u32, F>`. It became `Vec<F>` because the `small_field` machine needs two `F` for a register value.
Ideally, this would be `u32` as is the `MemoryState`, but the goldilocks machine actually uses the full `F` during its execution (e.g., for `to_signed`).

Reply:
I couldn't use the `BootloaderImpl` trait because it makes the crate dependencies circular, but maybe a `const` parameter might be possible... ill try